ДОМАШНЕЕ ЗАДАНИЕ №1

ВЫПОЛНИЛИ СТУДЕНТКИ БММ 183

МАРИЯ КРЕМНЕВА И УЛЬЯНА РИПП

In [30]:
import pandas as pd   
import numpy as np     
import matplotlib
import matplotlib.pyplot as plt
import seaborn as sns
plt.style.use('ggplot')
%matplotlib inline

In [31]:
#ЗАГРУЖАЕМ ФАЙЛЫ
ratings=pd.read_csv("ratings.csv") 
movies=pd.read_csv("movies.csv") 
tags=pd.read_csv("tags.csv")

In [32]:
ratings[['userId', 'movieId', 'rating']]

userId  movieId  rating
0            1        1     4.0
1            1        3     4.0
2            1        6     4.0
3            1       47     5.0
4            1       50     5.0
5            1       70     3.0
6            1      101     5.0
7            1      110     4.0
8            1      151     5.0
9            1      157     5.0
10           1      163     5.0
11           1      216     5.0
12           1      223     3.0
13           1      231     5.0
14           1      235     4.0
15           1      260     5.0
16           1      296     3.0
17           1      316     3.0
18           1      333     5.0
19           1      349     4.0
20           1      356     4.0
21           1      362     5.0
22           1      367     4.0
23           1      423     3.0
24           1      441     4.0
25           1      457     5.0
26           1      480     4.0
27           1      500     3.0
28           1      527     5.0
29           1      543     4.0
...        ...      ...     ...
100806     610   150401     3.0
100807     610   152077     4.0
100808     610   152081     4.0
100809     610   152372     3.5
100810     610   155064     3.5
100811     610   156371     5.0
100812     610   156726     4.5
100813     610   157296     4.0
100814     610   158238     5.0
100815     610   158721     3.5
100816     610   158872     3.5
100817     610   158956     3.0
100818     610   159093     3.0
100819     610   160080     3.0
100820     610   160341     2.5
100821     610   160527     4.5
100822     610   160571     3.0
100823     610   160836     3.0
100824     610   161582     4.0
100825     610   161634     4.0
100826     610   162350     3.5
100827     610   163937     3.5
100828     610   163981     3.5
100829     610   164179     5.0
100830     610   166528     4.0
100831     610   166534     4.0
100832     610   168248     5.0
100833     610   168250     5.0
100834     610   168252     5.0
100835     610   170875     3.0

[100836 rows x 3 columns]

In [33]:
movies['genres'].str.split("|")

0       [Adventure, Animation, Children, Comedy, Fantasy]
1                          [Adventure, Children, Fantasy]
2                                       [Comedy, Romance]
3                                [Comedy, Drama, Romance]
4                                                [Comedy]
5                               [Action, Crime, Thriller]
6                                       [Comedy, Romance]
7                                   [Adventure, Children]
8                                                [Action]
9                           [Action, Adventure, Thriller]
10                               [Comedy, Drama, Romance]
11                                       [Comedy, Horror]
12                       [Adventure, Animation, Children]
13                                                [Drama]
14                           [Action, Adventure, Romance]
15                                         [Crime, Drama]
16                                       [Drama, Romance]
17            

In [34]:
#ОБЪЕДИНЯЕМ ДВА ДАТАФРЕЙМА
movies = movies.merge(ratings, on = 'movieId', how = 'inner')
movies.head()

movieId             title                                       genres  \
0        1  Toy Story (1995)  Adventure|Animation|Children|Comedy|Fantasy   
1        1  Toy Story (1995)  Adventure|Animation|Children|Comedy|Fantasy   
2        1  Toy Story (1995)  Adventure|Animation|Children|Comedy|Fantasy   
3        1  Toy Story (1995)  Adventure|Animation|Children|Comedy|Fantasy   
4        1  Toy Story (1995)  Adventure|Animation|Children|Comedy|Fantasy   

   userId  rating   timestamp  
0       1     4.0   964982703  
1       5     4.0   847434962  
2       7     4.5  1106635946  
3      15     2.5  1510577970  
4      17     4.5  1305696483

In [35]:
#СОЗДАЕМ ДАТАСЕТ С ФИЛЬМАМИ,ЮЗЕРАМИ И ИХ ОЦЕНКАМИ
movies = movies[['userId','movieId','rating']]
movies.head()

userId  movieId  rating
0       1        1     4.0
1       5        1     4.0
2       7        1     4.5
3      15        1     2.5
4      17        1     4.5

In [38]:
conda install -c conda-forge scikit-surprise

Solving environment: ...working... done

# All requested packages already installed.


Note: you may need to restart the kernel to use updated packages.


In [39]:
from surprise import SVD
from surprise import Dataset, Reader
from surprise.model_selection import cross_validate

In [40]:
reader=Reader(rating_scale=(1,5))
data = Dataset.load_from_df(movies, reader)

In [42]:
algo = SVD()
cross_validate(algo, data, measures=['MSE'], cv=8, verbose=True)

Evaluating MSE of algorithm SVD on 8 split(s).

                  Fold 1  Fold 2  Fold 3  Fold 4  Fold 5  Fold 6  Fold 7  Fold 8  Mean    Std     
MSE (testset)     0.7624  0.7410  0.7623  0.7519  0.7389  0.7778  0.7499  0.7626  0.7559  0.0121  
Fit time          13.42   13.40   13.38   13.84   14.04   13.63   13.42   13.44   13.57   0.23    
Test time         0.30    0.30    0.31    0.44    0.31    0.30    0.36    0.31    0.33    0.05    


{'test_mse': array([0.76243066, 0.74103939, 0.76225174, 0.75190336, 0.73890422,
        0.77781944, 0.74992415, 0.76261871]),
 'fit_time': (13.423956632614136,
  13.39724850654602,
  13.379003047943115,
  13.835835456848145,
  14.044234991073608,
  13.627968311309814,
  13.419116258621216,
  13.438491582870483),
 'test_time': (0.3039684295654297,
  0.30304431915283203,
  0.3060035705566406,
  0.4437847137451172,
  0.305997371673584,
  0.30403590202331543,
  0.36499619483947754,
  0.3059389591217041)}

Итак, MSE = 0.7559 - это лучшее значение, которое нам удалось получить